### DISCUSSION TITLE
Using router with Tab Bar or Push Notification

### DISCUSSION THREAD

**Discussion by timofmax** [SwiftfulThinking/SwiftfulRouting#52] [Q&A]
Will be great to see description:
1)  how router should be implemented in App with TabBar.
As an example navigating from top of NavigationView in one screen to another tab with
2) how to open specific screen from push notification

**Comment by SwiftfulThinking** [comment]
1. You're app's onboarding / tabbar logic should be the same as it would be without using this package. Every app usually builds their own logic on how this will work specific to their product. Most commonly in SwiftUI apps I've seen one of these approaches:

### Switch between modules using conditional logic

##### Without SwiftfulRouting:
```Swift

ZStack {
    if isSignedIn {
        NavigationView {
            TabBarView()
        }
    }
    if !signedIn {
        NavigationView {
            OnboardingView()
        }
    }
}
```

##### With SwiftfulRouting:
```swift
ZStack {
    if isSignedIn {
        RouterView { _ in
            TabBarView()
        }
    }
    if !signedIn {
        RouterView { _ in
            OnboardingView()
        }
    }
}
```

### Show onboarding as a sheet or fullScreenCover

##### Without SwiftfulRouting:
```Swift

ZStack {
    NavigationView {
        TabBarView()
    }
    .onAppear {
        if !isSignedIn {
            showOnboardingView = true
        }
    }
    .sheet(isPresented: $showOnboardingView, content: {
        OnboardingView()
    })
}
```

##### With SwiftfulRouting:
```Swift
ZStack {
    RouterView { router in
        TabBarView()
            .onAppear {
                if !isSignedIn {
                    router.showScreen(.sheet) { _ in 
                         OnboardingView()
                    }
                }
            }
    }
}
```

2. Opening a deeplink would be approx. the same code as well, something like

```Swift
.onOpenURL { url in
     router.showScreen(.sheet) { _ in 
          MyScreen(url: url)
     }
}
```

**Comment by SwiftfulThinking** [comment]
If you have multiple tabs, you could alternatively do this so that each tab has it's own routing

```swift
ZStack {
    RouterView { _ in
        Tab1()
    }
    .tabBarItem {}
    
    RouterView { _ in
        Tab2()
    }
    .tabBarItem {}

    RouterView { _ in
        Tab3()
    }
    .tabBarItem {}

    RouterView { _ in
        Tab4()
    }
    .tabBarItem {}
}
```
